src/libxively/gateway/xi_gw_glue_layer.c

```diff
@@ -0,0 +1,123 @@
+/* Copyright (c) 2003-2018, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xi_gw_glue_layer.h>
+#include <xi_layer_api.h>
+#include <xively_mqtt.h>
+#include <xively.h>
+#include <xi_types.h>
+
+static void
+_xi_message_arrived_on_tunnel_callback( xi_context_handle_t in_context_handle,
+                                        xi_sub_call_type_t call_type,
+                                        const xi_sub_call_params_t* const params,
+                                        xi_state_t state,
+                                        void* user_data )
+{
+    printf( "--- %s ---, state: %d, user_data: %p\n", __FUNCTION__, state, user_data );
+
+    XI_UNUSED( in_context_handle );
+    XI_UNUSED( call_type );
+
+    if ( XI_MQTT_SUBSCRIPTION_SUCCESSFULL == state )
+    {
+        /* suback */
+    }
+    else if ( XI_STATE_OK == state )
+    {
+        /* PUBLISH arrived on tunnel */
+        xi_data_desc_t* data = xi_make_desc_from_buffer_copy(
+            params->message.temporary_payload_data,
+            params->message.temporary_payload_data_length );
+
+        XI_PROCESS_PULL_ON_THIS_LAYER( user_data, data, state );
+    }
```

Comment:
Also can't subacks have errors not just subscription successful?  Sorry, maybe not.. it's late friday night.

Reply:
Sure it can have suback failure. I've put there a todo_atigyi to cover this case as well.

Reply:
👍